### DISCUSSION TITLE
Wagmi useContractWrite Problem

### DISCUSSION THREAD

**Discussion by muhammad-rizwan-brainx** [wevm/wagmi#2416] [Q&A]
Hi, I have a problem with useContractWrite, I am approving tokens to a contract and then calling the function of contract that uses the approved tokens but useContractWrite is returning undefined and that is because it is unable to fetch latest allownce

**Comment by jxom** [comment]
I am not sure how to answer your question without more information and a code example. What functions are you calling? Do you have enough funds on your wallet? Are the functions being called in the right order? Is `useContractWrite` returning an error? What do you mean by “fetch latest allowance”? Particularly, a minimal reproduction via a https://new.wagmi.sh fork would be helpful.

**Reply by muhammad-rizwan-brainx** [reply]
These are my code snippets
For Approving
`export default function useApproveBUSD(amount: string) {
  const { config } = usePrepareContractWrite({
    address: BUSD_ADDRESS,
    abi: erc20ABI,
    functionName: "approve",
    args: [crowdSaleContractAddress, utils.parseEther(amount)],
  });

  const contract = useContractWrite(config);

  return contract;
}`

For Buying 
`export default function useBuyKelpBUSD(
  beneficiary: string,
  busdAmount: string
) {
  const { config } = usePrepareContractWrite({
    address: crowdSaleContractAddress,
    abi: crowdSaleABI,
    functionName: "buyTokensBUSD",
    args: [beneficiary, utils.parseEther(busdAmount)],
  });

  const contract = useContractWrite(config);

  return contract;
}`

I am calling buy right after approve is succeeded but useWriteContract continues giving undefined, if I put the gaslimit in overrides it gives writeAsync, but once i call writeAsynce and confirm from metamask transaction fails with error insufficient allowance, Although the allowance is available

**Reply by jxom** [reply]
You’ll probably need to wait for the approve transaction to have been processed on a block using `useWaitForTransaction`.

**Reply by muhammad-rizwan-brainx** [reply]
> I have tried using waitForTransaction but my issue is solved by passing gaslimit in args like this

`const { config } = usePrepareContractWrite({
    address: crowdSaleContractAddress,
    abi: crowdSaleABI,
    functionName: "buyTokensBUSD",
    chainId: 97,
    args: [beneficiary, utils.parseEther(busdAmount),
      {
        gasLimit: 100000,
      }],
  });
  const contract = useContractWrite(config);`
